### ISSUE TITLE
Nullable integers mapping issue

### ISSUE THREAD

**Issue by r-lopez** [jonwagner/Insight.Database#18]
I have the following Class definition:
public class Beer {
  public int Id { get; set; }
  public string Name { get; set; }
  public int? AlcoholPts { get; set; }
}

Executing my Insert stored procedure fails to map the AlcoholPts property, it always maps the stored procedure parameter value to 2046 (null as a value works properly). If I make the property not nullable the value when calling the stored procedure gets properly set (but I need to support nullable on this property). 

I am running VS 2012, SQL Server 2012, Insight.Database v2.0.1 for target framework net45.

Has anyone faced this issue before? Does anyone know about a solution or a way around without having to create two different class definition for beer? 

Thanks!

**Comment by chdev77** [comment]
Jon, I just upgraded from 4.2.10.0 to 5.2.5.0 and I'm getting this exception

Exception thrown: 'System.InvalidCastException' in mscorlib.dll

Additional information: Invalid cast from 'System.Int32' to 'System.Nullable`1[[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]'.

If I make the type non nullable everything works, but I cannot refactor everything to fix this.  Any ideas on how to fix or a work around.  I have several layers and the main is MVC running .NET 4.5.  Changing the platform target from Auto to x86 on all projects does nothing.

**Comment by jonwagner** [comment]
Can you post some code to reproduce the problem?

**Comment by chdev77** [comment]
public ClaimSearchResults Search(IClaimSearchCriteria criteria)
        {
            using (var db = Databases.CaseManagement)
            {
                criteria.ProjectId = SecurityContext.ProjectId;
                var parameters = criteria.Map<ClaimSearchParameters>();

                var results = db.Query("dbo.ClaimSearch", parameters, Query.Returns<ClaimSearchResult>()
                    .ThenChildren(OneToOne<Debtor>.Records, id: c => c.LiabilityId.Value, into: (c, cd) => PopulateDebtorFields(c, cd))
                    .ThenChildren(Some<Tag>.Records, id: c => c.LiabilityId.Value, into: (c, ct) => c.Tags = ct)
                    .ThenChildren(Some<CategoryType>.Records, id: c => c.LiabilityId.Value, into: (c, ct) => c.ClaimCategories = ct));

                return new ClaimSearchResults
                {
                    Items = results.ToArray(),
                    TotalRowCount = parameters.TotalRowCount ?? 0
                };
            }
        }


on the LiabilityId.Value is the nullable I didn't used to need the .Value and now I do.  That fixed the problem here, but now I get....

using net 4.5.2  get Exception thrown: 'System.Security.VerificationException' in System.Core.dll Additional information: Operation could destabilize the runtime.

on the next method...
        public Contacts Get(ContactCriteria criteria)
        {
            using (var db = Databases.CaseManagement)
            {
                criteria.ProjectId = SecurityContext.ProjectId;
                var parameters = criteria.Map<ContactSearchParameters>();

                var results = db.Query("dbo.ContactSearch", parameters, Query
                    .Returns<Contact>()
                    .ThenChildren(Some<Address>.Records, id: a => a.ContactId.Value, into: (c, a) => c.Addresses = a.ToArray().AddPrimaryIfNotExists())
                    .ThenChildren(Some<PhoneNumber>.Records, id: a => a.ContactId, into: (c, p) => c.PhoneNumbers = p.ToArray())
                    .ThenChildren(Some<EmailAddress>.Records, id: a => a.ContactId.Value, into: (c, e) => c.EmailAddresses = e.ToArray())
                    .ThenChildren(Some<Tag>.Records, id: a => a.ContactId.Value, into: (c, t) => c.Tags = t.ToArray())
                    .ThenChildren(Some<ListContact>.Records, id: a => a.ContactId.Value, into: (c, a) => c.Lists = a.ToList()));

                return new Contacts
                {
                    Items = results.ToArray(),
                    TotalRowCount = parameters.TotalRowCount ?? 0
                };
            }
        }


let me know what code you need because I have a lot

**Comment by chdev77** [comment]
If I downgrade to Insight.Database 5.0.0 I still get the Casting error and have to change my nullable Id's to Id.Value but I don't get the exception...

Exception thrown: 'System.Security.VerificationException' in System.Core.dll Additional information: Operation could destabilize the runtime.

**Comment by chdev77** [comment]
Except for when the nullable is actually null...I guess I'll have to write an extension to handle this problem unless you know of how to resolve the initial problem.

![image](https://cloud.githubusercontent.com/assets/7691322/7923776/c409e47e-087b-11e5-9a70-d0a55d21f9ee.png)

**Comment by chdev77** [comment]
I forked Insight and pulled the projects into my solution and referenced the projects from mine.  The exception is in ChildMapper.cs line 79, its trying to convert from (int) 89 to (int?) 89 and causing the casting exception.  
![image](https://cloud.githubusercontent.com/assets/7691322/7927417/5337adc4-08a5-11e5-99e8-f9ccca44752c.png)

**Comment by chdev77** [comment]
I downgraded to version 4.2.5 and this solve my problems for now.  I also tried version 5.0.0 and still had the casting problems.

**Comment by jonwagner** [comment]
Thanks for looking into this. I’m going to try to do a round of Insight upgrades next week.

From:  chdev77
Reply-To:  "jonwagner/Insight.Database"
Date:  Tuesday, June 2, 2015 at 9:51 AM
To:  "jonwagner/Insight.Database"
Cc:  Jon Wagner
Subject:  Re: [Insight.Database] Nullable integers mapping issue (#18)


I downgraded to version 4.2.5 and this solve my problems for now.  I also tried version 5.0.0 and still had the casting problems.
